codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -438,9 +455,176 @@ public void operationComplete(ChannelFuture future) throws Exception {
+            super(ctx, stream, padding, endOfStream, promise);
+            this.data = data;
+            size = dataLength + padding;
+        }
+
+        @Override
+        public long size() {
+            return size;
+        }
+
+        /**
+         * Return {@link Integer#MAX_VALUE} if the actual size is larger than
+         * {@link Integer#MAX_VALUE}.
+         */
+        protected abstract int writableData();
+
+        protected abstract T sliceData(int length);
+
+        protected abstract void writeData(T toWrite, int padding, boolean endStream,
+                ChannelPromise writePromise);
+
+        private ChannelPromise selectPromise(int bytesWritten) {
+            if (size == bytesWritten && !promise.isVoid()) {
+                // Can use the original promise if it's the last write
+                return promise;
+            } else {
+                // Create a new promise and listen to it for failure
+                ChannelPromise newPromise = ctx.newPromise();
+                newPromise.addListener(this);
+                return newPromise;
+            }
+        }
+
+        private int writePaddingOnly(int allowedFrameSize, int bytesWritten) {
+            int writablePadding = Math.min(allowedFrameSize, padding);
+            padding -= writablePadding;
+            bytesWritten += writablePadding;
+            ChannelPromise writePromise = selectPromise(bytesWritten);
+            frameWriter().writeData(ctx, stream.id(), Unpooled.EMPTY_BUFFER, writablePadding,
+                    size == bytesWritten && endOfStream, writePromise);
+            return bytesWritten;
+        }
+
+        @SuppressWarnings("unchecked")
+        private int writeFrame(int allowedFrameSize, int bytesWritten) {
+            T toWrite;
+            int writableData = writableData();
+            // Let data consume the frame before padding.
+            if (writableData > allowedFrameSize) {
+                writableData = allowedFrameSize;
+                toWrite = (T) sliceData(allowedFrameSize).retain();
+            } else {
+                // We're going to write the full buffer which will cause it to be released. Set data
+                // to null to indicate subsequent writes(if any) that we only need to write padding.
+                toWrite = data;
+                data = null;
+            }
+            int writablePadding = Math.min(allowedFrameSize - writableData, padding);
+            padding -= writablePadding;
+            bytesWritten += writableData + writablePadding;
+            ChannelPromise writePromise = selectPromise(bytesWritten);
+            writeData(toWrite, writablePadding, size == bytesWritten && endOfStream, writePromise);
+            return bytesWritten;
+        }
+
+        @Override
+        public void write(int allowedBytes) {
+            int bytesWritten = 0;
+            if (data == null || (allowedBytes == 0 && size != 0)) {
```

Comment:
I agree with @nmittler's above comment.  Either try to unify them or keep them separate.

It would also be good to unify these two methods to make sure both cases work.  For example in this case you are setting `data = null` to signify you are done writing data, and may have padding left...but here you are checking `data == null` and exiting before checking if any padding needs to be written.  The `ByteBuf` accounts for this by only setting `data = null` on release, and `EMPTY_BUF` if the write process exhausted the available data in the buffer.